benchmarks/infill_density/infill_ascii_data.cc

```diff
@@ -0,0 +1,253 @@
+
+        /**
+         * Declare the parameters this class takes through input files. The
+         * default implementation of this function does not describe any
+         * parameters. Consequently, derived classes do not have to overload
+         * this function if they do not take any runtime parameters.
+         */
+        static
+        void
+        declare_parameters (ParameterHandler &prm);
+
+        /**
+         * Read the parameters this class declares from the parameter file.
+         * The default implementation of this function does not read any
+         * parameters. Consequently, derived classes do not have to overload
+         * this function if they do not take any runtime parameters.
+         */
+        void
+        parse_parameters (ParameterHandler &prm) override;
+
+      private:
+        types::boundary_id surface_boundary_id;
+        double rock_density;
+        double sediment_density;
+        double crustal_density;
+        double rock_infill_height;
+    };
+
+    template <int dim>
+    InfillAsciiData<dim>::InfillAsciiData ()
+      :
+      surface_boundary_id(numbers::invalid_unsigned_int)
+    {}
+
+
+    template <int dim>
+    void
+    InfillAsciiData<dim>::initialize ()
+    {
+      surface_boundary_id = this->get_geometry_model().translate_symbolic_boundary_name_to_id("top");
+      std::set<types::boundary_id> surface_boundary_set;
+      surface_boundary_set.insert(surface_boundary_id);
+
+      // The input ascii table contains one data column (LAB depths(m)) in addition to the coordinate columns.
+      Utilities::AsciiDataBoundary<dim>::initialize(surface_boundary_set,
+                                                    1);
+    }
+
+
+    template <int dim>
+    Tensor<1,dim>
+    InfillAsciiData<dim>::
+    boundary_traction (const types::boundary_id surface_boundary_id,
+                       const Point<dim> &position,
+                       const Tensor<1,dim> &normal_vector) const
+    {
+      const double load = Utilities::AsciiDataBoundary<dim>::get_data_component(surface_boundary_id,
+                                                                                position,
+                                                                                0);
+      Tensor<1, dim> traction = normal_vector;
+      const double gravity_norm = this->get_gravity_model().gravity_vector(position).norm();
+      \
+      const double elevation = this->get_geometry_model().height_above_reference_surface(position);
+
+      // Check if the load at the current point has a height which exceeds the
+      // 'infill height'. This is useful for more complicated topographic loads
+      // when its harder to define the precise location of a volcanic edifice,
+      // for example. If the load does exceed the rock_infill_height, the infill
+      // material has the density of rock, otherwise it has sediment density.
+      if (load >= rock_infill_height*gravity_norm*rock_density)
```

Comment:
I am trying to understand this but I am not quite sure I do. Can you explain a little bit more the reasoning behind how the load from the ascii file is modified and what the assumptions are? Here is what I think you do: the idea is that you know the actual height of the volcano (but not the actual load), so you have to modify the load based on what the actual topography in the model is? If it's negative, you increase the load because you would have infill in the depression the load creates. And if it's positive, you decrease the load because some of it is already accounted for by the positive surface topography?

But I don't really understand the infill_height part. Why does this depend on the load, and not on the elevation?
Why does it use the `rock_density` rather than `sediment_density`? In the current file, it is zero, so for the benchmark this does not make a difference, but I imagine for an application model is does. So it would be good to explain this (and the bast place might be the class description, or the parameter descriptions).

Reply:
I updated the documentation but maybe it would be best if I also included another .png showing how this works? I'm finding it hard to explain in just text alone.

Reply:
The figure definitely helps. I don't have any suggestions to further improve the documentation above, and think it is in good shape.